orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/providers/gce/GoogleImageTagger.java

```diff
@@ -0,0 +1,192 @@
+    StageData stageData = (StageData) stage.mapTo(StageData.class);
+
+    Collection<MatchedImage> matchedImages = findImages(stageData.imageNames, stage);
+
+    stageData.imageNames = matchedImages.stream()
+      .map(matchedImage -> matchedImage.imageName)
+      .collect(Collectors.toSet());
+
+    // Built-in tags are not updatable
+    Map<String, String> tags = stageData.tags.entrySet().stream()
+      .filter(entry -> !BUILT_IN_TAGS.contains(entry.getKey()))
+      .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    List<Image> targetImages = new ArrayList<>();
+    Map<String, Object> originalTags = new HashMap<>();
+    List<Map<String, Map>> operations = new ArrayList<>();
+
+    for (MatchedImage matchedImage : matchedImages) {
+      Image targetImage = new Image(
+        matchedImage.imageName,
+        getCredentials(stage),
+        null,
+        tags
+      );
+      targetImages.add(targetImage);
+
+      log.info(format("Tagging '%s' with '%s' (executionId: %s)", targetImage.imageName, targetImage.tags, stage.getExecution().getId()));
+
+      // Update the tags on the image
+      operations.add(
+        ImmutableMap.<String, Map>builder()
+          .put(OPERATION, ImmutableMap.builder()
+            .put("imageName", targetImage.imageName)
+            .put("tags", targetImage.tags)
+            .put("credentials", targetImage.account)
+            .build()
+          ).build()
+      );
+
+      originalTags.put(matchedImage.imageName, matchedImage.tags);
+    }
+
+    Map<String, Object> extraOutput = objectMapper.convertValue(stageData, Map.class);
+    extraOutput.put("targets", targetImages);
+    extraOutput.put("originalTags", originalTags);
+
+    return new ImageTagger.OperationContext(operations, extraOutput);
+  }
+
+  /**
+   * Return true iff the tags on the current machine image match the desired.
+   */
+  @Override
+  public boolean areImagesTagged(Collection<Image> targetImages, Stage stage) {
+    Collection<MatchedImage> matchedImages = findImages(
+      targetImages.stream().map(targetImage -> targetImage.imageName).collect(Collectors.toSet()),
+      stage
+    );
+
+    AtomicBoolean isUpserted = new AtomicBoolean(true);
+
+    for (Image targetImage : targetImages) {
+      MatchedImage matchedImage = matchedImages.stream()
+        .filter(m -> m.imageName.equals(targetImage.imageName))
+        .findFirst()
+        .orElse(null);
+
+      if (matchedImage == null) {
+        isUpserted.set(false);
+      } else {
+        targetImage.tags.entrySet().forEach(entry -> {
+          // assert tag equality
+          isUpserted.set(isUpserted.get() && entry.getValue().equals(matchedImage.tags.get(entry.getKey())));
+        });
+      }
+    }
+
+    return isUpserted.get();
+  }
+
+  @Override
+  public String getCloudProvider() {
+    return "gce";
+  }
+
+  private Collection<MatchedImage> findImages(Collection<String> imageNames, Stage stage) {
```

Comment:
If I'm reading this correctly, it'll throw whenever it doesn't find a desired image in `imageNames`. Why bother with [this check](https://github.com/spinnaker/orca/pull/1106/files#diff-b832c3399f681baf48877ccde1287963R118) if `targetImages` contains all images in `matchedImages` given that it hasn't thrown at that point?